### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
@mbudiu-vmw, just got back into this today (haven't done too much, but thanks again for the powerpoint).

### CODE REVIEW

File: backends/ebpf/ebpfType.h [hunk 1]
```diff
@@ -146,6 +146,30 @@ class EBPFEnumType : public EBPFType, public EBPF::IHasWidth {
     const IR::Type_Enum* getType() const { return type->to<IR::Type_Enum>(); }
 };
 
+class EBPFStackType : public EBPFType, public IHasWidth {
+    class EBPFHdr {
```
Comment (mihaibudiu) [inline backends/ebpf/ebpfType.h]: I don't think you need this class.

File: backends/ebpf/ebpfType.h [hunk 2]
```diff
@@ -146,6 +146,30 @@ class EBPFEnumType : public EBPFType, public EBPF::IHasWidth {
...
+        cstring comment;
+        EBPFType* type;
+        const IR::Type_Header* hdr;
+
+    EBPFHdr(EBPFType* type, const IR::Type_Header* hdr, cstring comment = nullptr) :
+        comment(comment), type(type), hdr(hdr) {}
+    };
+
+ public:
+     std::vector<EBPFHdr*> hdrs;
```
Comment (mihaibudiu) [inline backends/ebpf/ebpfType.h]: You also probably don't need this vector, you just need to know the element type to generate C code for an array of such elements.

**Review by mihaibudiu — COMMENTED** [review]
I am planning to overhauls the C code generation from ebpf soon; the plan is to keep headers packed and unpack data only when necessary.

However, I expect to start working on this problem only in January.

File: backends/ebpf/ebpfType.h (hunk 1 above)
Comment (zeeshanlakhani) [reply backends/ebpf/ebpfType.h]: yep, doing a rewrite now @mbudiu-vmw, will push an update somewhat later.

**Comment by zeeshanlakhani** [comment]
@mbudiu-vmw, 

took a better approach. The .c and .h generated-files are started to make more sense now in each of the steps. However, think that the next issue may be how to handle `next` and other operations. Any feedback/guidance would be helpful. 

Also, glad to know you'll be working on ebpf->C generation in January. Even if we can help get this basically working, for now, that's good for us too :).

**Comment by mihaibudiu** [comment]
The handling of next is defined in the P4-16 spec. You can implement it by creating another local variable to hold the index of the next element, which you increment every time you compile an extract(array.next); So extract(array.next) becomes `extract(arrayrepr[index]); index++;`. array.last becomes `arrayrepr[index-1]`.

**Comment by mihaibudiu** [comment]
Please be aware that the current implementation is very inconsistent with respect to the byte ordering of the C representation. This is a problem that we'll have to fix; we are writing a spec about how this should be done, which will be the blueprint for our reimplementation.

**Comment by zeeshanlakhani** [comment]
@mbudiu-vmw, I just got back to this (holidays and other stuffz). If you do get anywhere on ebpf as a whole, we'd love to drop trial'ing through this :). Nonetheless, anywhere good to look or any recommendations for unrolling each state of `.next`, if the transition determines that we should revisit the parser state again, since we're talking `ebpf` here. Or, should I not worry about unrolling in the case you mentioned (though I think I should :) ). Thanks!

**Comment by mihaibudiu** [comment]
I actually plan to rewrite a significant portion of the ebpf code generation to be more efficient.
Probably starting next week.
There is a loop unrolling pass, but it is commented-out and has rotted; I will have to fix it too.
It is part of the mid-end.

**Comment by zeeshanlakhani** [comment]
Great @mbudiu-vmw, update us and we'll review too along the way. Thanks!

**Comment by mihaibudiu** [comment]
I will probably create a separate branch.
The plan is to keep headers packed and structures unpacked.

**Comment by zeeshanlakhani** [comment]
Great, we'll follow progress.

**Pull request closed without merging**